File: ios/apps/AutoTester/AutoTester/testCases/AnimationDelegateTestCase.swift
//
//  AnimationDelegateTestCase.swift
//  AutoTester
//
//  Created by Ranen Ghosh on 2016-11-29.
//  Copyright 2016-2022 mousebird consulting.
//

import UIKit
import WhirlyGlobe

class AnimationDelegateTestCase: MaplyTestCase {
    override init() {
        super.init()
        
        self.name = "Animation Delegate"
        self.implementations = [.globe, .map]
    }
    
    var baseCase = StamenWatercolorRemote()
    let initPos = MaplyCoordinateMakeWithDegrees(12.5037997, 41.893988)
    let pos = MaplyCoordinateMakeWithDegrees(-0.1275, 51.507222)

    override func setUpWithGlobe(_ globeVC: WhirlyGlobeViewController) {
        baseCase.setUpWithGlobe(globeVC)
        
        let minZ: Float = 0.001 //globeVC.getZoomLimitsMin()
        let maxZ: Float = 2.0 // globeVC.getZoomLimitsMax()

        globeVC.setPosition(initPos)    // this doesn't seem to work?
        globeVC.height = maxZ
        globeVC.keepNorthUp = false
        globeVC.animate(toPosition:pos, height:minZ, heading:Float.pi/4.0, time:9.0);

        // Do it again with the old, linear zoom and keep-north-up
        Timer.scheduledTimer(withTimeInterval: 10, repeats: false) { [self] _ in
            globeVC.keepNorthUp = true
            globeVC.animationZoomEasing = { (z0,z1,t) in z0 + (z1 - z0) * t }
            globeVC.animate(toPosition: initPos, height: maxZ, heading: Float.pi/4.0, time: 5.0)
        }
        Timer.scheduledTimer(withTimeInterval: 15, repeats: false) { [self] _ in
            globeVC.animate(toPosition: pos, height: minZ, heading: 0.0, time: 5.0)
        }

        // And again, with flair!
        Timer.scheduledTimer(withTimeInterval: 22, repeats: false) { [self] _ in
            globeVC.animationZoomEasing = easeOut
            globeVC.animate(toPosition: initPos, height: maxZ, heading: 0.0, time: 5.0)
        }
        Timer.scheduledTimer(withTimeInterval: 28, repeats: false) { [self] _ in
            globeVC.animationZoomEasing = easeIn
            globeVC.animate(toPosition: pos, height: minZ, heading: 0.0, time: 5.0)
        }
    }
    
    override func setUpWithMap(_ mapVC: MaplyViewController) {
        baseCase.setUpWithMap(mapVC)

        let minZ = mapVC.getMinZoom()
        let maxZ = Float(2.0)

        // Start high, zoom in
        mapVC.setPosition(initPos)
        mapVC.height = maxZ;
        mapVC.animate(toPosition:pos, height:minZ, heading:Float.pi/4.0, time:9.0);

        // Back out, and in again, with the old, linear zoom
        Timer.scheduledTimer(withTimeInterval: 10, repeats: false) { [self] _ in
            mapVC.animationZoomEasing = { (z0,z1,t) in z0 + (z1 - z0) * t }
            mapVC.animate(toPosition: initPos, height: maxZ, heading: 0.0, time: 5.0)
        }
        Timer.scheduledTimer(withTimeInterval: 15, repeats: false) { [self] _ in
            mapVC.animate(toPosition: pos, height: minZ, heading: 0.0, time: 5.0)
        }

        // And again, with flair!
        Timer.scheduledTimer(withTimeInterval: 22, repeats: false) { [self] _ in
            mapVC.animationZoomEasing = easeOut
            mapVC.animate(toPosition: initPos, height: maxZ, heading: 0.0, time: 5.0)
        }
        Timer.scheduledTimer(withTimeInterval: 28, repeats: false) { [self] _ in
            mapVC.animationZoomEasing = easeIn
            mapVC.animate(toPosition: pos, height: minZ, heading: 0.0, time: 5.0)
        }
    }

    private let easeIn = linearize({ (z0,z1,t) in z0+(z1-z0)*easeInBack(t) })
    private let easeOut = linearize({ (z0,z1,t) in z0+(z1-z0)*(1-easeInBack(1-t)) })

    private static func linearize(_ f: @escaping (Double,Double,Double)->Double) -> ((Double,Double,Double)->Double) {
        return { (z0,z1,t) in exp(f(log(z0),log(z1),t)) }
    }
    // https://easings.net/#easeInBack
    private static func easeInBack(_ t: Double) -> Double {
        let c = 1.70158;
        return ((c + 1) * t - c) * t * t;
    }
}

File: android/apps/AutoTesterAndroid/app/src/main/java/com/mousebirdconsulting/autotester/TestCases/AnimationDelegateTestCase.kt
/*  AnimationDelegateTestCase.kt
 *  WhirlyGlobeLib
 *
 *  Created by Tim Sylvester on 9 Feb 2021.
 *  Copyright 2021 mousebird consulting
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package com.mousebirdconsulting.autotester.TestCases

import android.app.Activity
import android.util.Log
import android.widget.Toast
import com.mousebird.maply.*
import com.mousebirdconsulting.autotester.Framework.MaplyTestCase
import kotlin.math.exp
import kotlin.math.log

class AnimationDelegateTestCase(activity: Activity) :
        MaplyTestCase(activity, "Animation Delegate", TestExecutionImplementation.Both) {
    
    private val initLoc = Point2d.FromDegrees(12.5037997, 41.893988)
    private val loc = Point2d.FromDegrees(0.5508867, 51.5203505)
    private val fromHeight = 1.5
    private val toHeight = 0.01

    override fun setUpWithGlobe(vc: GlobeController): Boolean {
        if (!baseCase.setUpWithGlobe(vc)) {
            return false
        }
        vc.apply {
            keepNorthUp = false
            positionGeo = Point3d(initLoc.x, initLoc.y, fromHeight)
            heading = degToRad(0.0)
        }
        script(vc, sequenceOf(
            // animate
            ScriptItem(10.0, null) { vc.animatePositionGeo(loc, toHeight, degToRad(45.0), 9.0) },
        
            // Animate heading only
            ScriptItem(2.0, "Heading Only") { vc.animatePositionGeo(loc, toHeight, degToRad(-45.0), 2.0) },
            // and back
            ScriptItem(3.0, null) { vc.animatePositionGeo(loc, toHeight, degToRad(0.0), 2.0) },
        
            // repeat with `keepNorthUp` set
            ScriptItem(5.0, "Keep North Up") {
                vc.keepNorthUp = true
                vc.animatePositionGeo(initLoc, fromHeight, degToRad(45.0), 5.0)
            },
            ScriptItem(6.0, null) { vc.animatePositionGeo(loc, toHeight, degToRad(45.0), 5.0) },
        
            // Again with old, linear zoom scale
            ScriptItem(5.0, "Linear Zoom") {
                vc.zoomAnimationEasing = BaseController.ZoomAnimationEasing { z0, z1, t -> z0 + (z1 - z0) * t }
                vc.animatePositionGeo(loc, fromHeight, 0.0, 5.0)
            },
            ScriptItem(6.0, null) { vc.animatePositionGeo(loc, toHeight, 0.0, 5.0) },
        
            // Again, with flair!
            ScriptItem(6.0, "Flair") {
                vc.zoomAnimationEasing = BaseController.ZoomAnimationEasing(easeInOnly)
                vc.animatePositionGeo(loc, fromHeight, 0.0, 5.0)
            },
            ScriptItem(6.0, null) { vc.animatePositionGeo(loc, toHeight, 0.0, 5.0) },
        ))
        return true
    }
    
    override fun setUpWithMap(vc: MapController): Boolean {
        if (!baseCase.setUpWithMap(vc)) {
            return false
        }
        vc.apply {
            positionGeo = Point3d(initLoc.x, initLoc.y, fromHeight)
            height = fromHeight
            // Note that, unlike `GlobeViewController.keepNorthUp`, this option only affects
            // gestures, not programmatic changes through `animatePositionGeo` and friends
            allowRotateGesture = false
        }
        script(vc, sequenceOf(
            // animate
            ScriptItem(10.0, null) { vc.animatePositionGeo(loc, toHeight, degToRad(45.0), 9.0) },
        
            // Animate heading only
            ScriptItem(2.0, "Heading Only") { vc.animatePositionGeo(loc, toHeight, degToRad(-45.0), 2.0) },
            // and back
            ScriptItem(3.0, null) { vc.animatePositionGeo(loc, toHeight, degToRad(0.0), 2.0) },
        
            // Again with old, linear zoom scale
            ScriptItem(5.0, "Linear Zoom") {
                vc.zoomAnimationEasing = BaseController.ZoomAnimationEasing { z0, z1, t -> z0 + (z1 - z0) * t }
                vc.animatePositionGeo(loc, fromHeight, 0.0, 5.0)
            },
            ScriptItem(6.0, null) { vc.animatePositionGeo(loc, toHeight, 0.0, 5.0) },
        
            // Again, with flair!
            ScriptItem(6.0, "Flair") {
                vc.zoomAnimationEasing = BaseController.ZoomAnimationEasing(easeInOnly)
                // We have to target a lower zoom level for now because the Android
                // flat map doesn't handle zooming out out beyond about 2 well.
                vc.animatePositionGeo(loc, 1.0, 0.0, 5.0)
            },
            ScriptItem(6.0, null) { vc.animatePositionGeo(loc, toHeight, 0.0, 5.0) },
        ))
        return true
    }

    override fun shutdown() {
        baseCase.shutdown()
        super.shutdown()
    }
    
    private data class ScriptItem(
        val duration: Double,
        val toast: String?,
        val run: ()->Unit)
    private fun script(vc: BaseController, items: Sequence<ScriptItem>) {
        var t = 0.0
        items.forEach { (duration, toastStr, run) ->
            vc.addPostSurfaceRunnable({ toast(toastStr) }, (t * 1000.0).toLong())
            vc.addPostSurfaceRunnable(run, (t * 1000.0).toLong())
            t += duration
        }
    }
    
    // https://easings.net/#easeInBack
    private fun easeInBack(t: Double): Double {
        val c = 1.70158
        return ((c + 1) * t - c) * t * t;
    }
    private fun linearize(f: ((Double,Double,Double)->Double)): ((Double,Double,Double)->Double) {
        return { z0,z1,t -> exp(f(log(z0,Math.E),log(z1,Math.E),t)) }
    }
    // Apply easing only on zooming in to avoid going negative
    private val easeInOnly = linearize { z0,z1,t ->
        z0+(z1-z0)*(if (z0>z1) easeInBack(t) else (1-easeInBack(1-t)))
    }
    
    private var baseCase = VectorsTestCase(activity)
    
    private var prevToast: Toast? = null
    private fun toast(s: String?) {
        activity?.runOnUiThread {
            prevToast?.cancel()
            prevToast = s?.let {
                activity?.applicationContext?.let { c ->
                    Toast.makeText(c, s, Toast.LENGTH_SHORT).also {
                        it.show()
                    }
                }
            }
        }
    }
}
